File: android/src/main/kotlin/com/zeticai/zetic_mlange_flutter/wrapper/core/model/ZeticMLangeModelFlutterPlugin.kt
package com.zeticai.zetic_mlange_flutter.wrapper.core.model

import android.content.Context
import com.zeticai.mlange.core.error.ZeticMLangeException
import com.zeticai.mlange.core.model.ZeticMLangeModel
import com.zeticai.zetic_mlange_flutter.core.BaseFlutterPlugin

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.nio.ByteBuffer

class ZeticMLangeModelFlutterPlugin(
    flutterPluginBinding: FlutterPlugin.FlutterPluginBinding,
    channelStr: String,
    private val context: Context = flutterPluginBinding.applicationContext
) : BaseFlutterPlugin<ZeticMLangeModel>(flutterPluginBinding, channelStr) {
    private fun create(instanceId: String, tokenKey: String, modelKey: String, result: Result) {
        Thread {
            try {
                val model = ZeticMLangeModel(context, tokenKey, modelKey)
                createInstance(instanceId, model)
                result.success(null)
            } catch (e: Exception) {
                result.error(
                    "CREATE_FAIL",
                    "Failed to create ZeticMLangeModel: ${e.message}",
                    null
                )
            }
        }.start()
    }

    private fun run(instanceId: String, inputs: List<ByteArray>, result: Result) {
        val model = getInstance(instanceId, result)

        try {
            val byteBuffers = inputs.map { ByteBuffer.wrap(it) }.toTypedArray()
            model.run(byteBuffers)
            result.success(null)
        } catch (e: ZeticMLangeException) {
            result.error(
                "RUN_ERROR",
                e.message,
                null
            )
        }
    }

    private fun getOutputDataArray(instanceId: String, result: Result) {
        val model = getInstance(instanceId, result)

        val outputBuffers = model.getOutputBuffers()
        val outputList = outputBuffers.map {
            val byteArray = ByteArray(it.remaining())
            it.get(byteArray)
            byteArray
        }

        result.success(outputList)
    }

    private fun deinit(instanceId: String, result: Result) {
        removeInstance(instanceId)?.let {
            try {
                it.deinit()
            } catch (e: Exception) {
                result.error(
                    "DEINIT_ERROR",
                    e.message,
                    null
                )
            }
        }
        result.success(null)
    }
}

File: ios/Classes/wrapper/ZeticMLangeModelFlutterPlugin.swift
import Flutter
import UIKit
import ZeticMLange

public class ZeticMLangeModelFlutterPlugin: BaseFlutterPlugin {
    let instances = MultipleInstances<ZeticMLangeModel>()
    
    public override func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "create":
            guard
                let args = call.arguments as? [String: Any],
                let instanceId = args["instanceId"] as? String,
                let tokenKey = args["tokenKey"] as? String,
                let modelKey = args["modelKey"] as? String
            else {
                result(FlutterError(code: "ARGUMENT_ERROR", message: "Invalid arguments for create", details: nil))
                return
            }
            create(instanceId: instanceId, tokenKey: tokenKey, modelKey: modelKey, result: result)
            
        case "run":
            guard
                let args = call.arguments as? [String: Any],
                let instanceId = args["instanceId"] as? String,
                let inputData = args["inputs"] as? [FlutterStandardTypedData]
            else {
                result(FlutterError(code: "ARGUMENT_ERROR", message: "Invalid arguments for run", details: nil))
                return
            }
            run(instanceId: instanceId, inputs: inputData, result: result)

        case "getOutputDataArray":
            guard
                let args = call.arguments as? [String: Any],
                let instanceId = args["instanceId"] as? String
            else {
                result(FlutterError(code: "ARGUMENT_ERROR", message: "Invalid arguments for getOutputDataArray", details: nil))
                return
            }
            getOutputDataArray(instanceId: instanceId, result: result)
            
        case "deinit":
            guard
                let args = call.arguments as? [String: Any],
                let instanceId = args["instanceId"] as? String
            else {
                result(FlutterError(code: "ARGUMENT_ERROR", message: "Invalid arguments for deinit", details: nil))
                return
            }
            _deinit(instanceId: instanceId, result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func create(instanceId: String, tokenKey: String, modelKey: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .background).async {
            do {
                let model = try ZeticMLangeModel(throws: tokenKey, throws: modelKey)
                
                self.instances.createInstance(instanceId: instanceId, instance: model)
                
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(
                        FlutterError(
                            code: "CREATE_FAIL",
                            message: "Failed to create ZeticMLangeModel: \(error.localizedDescription)",
                            details: nil
                        )
                    )
                }
            }
        }
    }

    func run(instanceId: String, inputs: [FlutterStandardTypedData], result: @escaping FlutterResult) {
        let model = instances.getInstance(instanceId: instanceId, result: result)
        do {
            let dataInputs = inputs.map { $0.data }
            try model.run(dataInputs)
            result(nil)
        } catch let e as ZeticMLangeError {
            result(FlutterError(
                code: "RUN_ERROR",
                message: e.localizedDescription,
                details: nil
            ))
        } catch {
            result(FlutterError(
                code: "RUN_ERROR",
                message: error.localizedDescription,
                details: nil
            ))
        }
    }

    func getOutputDataArray(instanceId: String, result: @escaping FlutterResult) {
        let model = instances.getInstance(instanceId: instanceId, result: result)
        let outputBuffers: [Data] = model.getOutputDataArray()
        let outputList: [FlutterStandardTypedData] = outputBuffers.map { FlutterStandardTypedData(bytes: $0) }
    
        result(outputList)
    }

    func _deinit(instanceId: String, result: @escaping FlutterResult) {
        instances.removeInstance(instanceId: instanceId)
        result(nil)
    }
}
